integrations/access/servicenow/app.go

```diff
@@ -0,0 +1,591 @@
+// createIncident posts an incident with request information.
+func (a *App) createIncident(ctx context.Context, serviceID, reqID string, reqData RequestData) error {
+	data, err := a.servicenow.CreateIncident(ctx, reqID, reqData)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	ctx, log := logger.WithField(ctx, "servicenow_incident_id", data.IncidentID)
+	log.Info("Successfully created Servicenow incident")
+
+	// Save servicenow incident info in plugin data.
+	_, err = a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		var pluginData PluginData
+		if existing != nil {
+			pluginData = *existing
+		} else {
+			// It must be impossible but lets handle it just in case.
+			pluginData = PluginData{RequestData: reqData}
+		}
+		pluginData.ServicenowData = ServicenowData{IncidentID: data.IncidentID}
+		return pluginData, true
+	})
+	return trace.Wrap(err)
+}
+
+// postReviewNotes posts incident notes about new reviews appeared for request.
+func (a *App) postReviewNotes(ctx context.Context, reqID string, reqReviews []types.AccessReview) error {
+	var oldCount int
+	var data ServicenowData
+
+	// Increase the review counter in plugin data.
+	ok, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		if existing == nil {
+			return PluginData{}, false
+		}
+
+		if data = existing.ServicenowData; data.IncidentID == "" {
+			return PluginData{}, false
+		}
+
+		count := len(reqReviews)
+		if oldCount = existing.ReviewsCount; oldCount >= count {
+			return PluginData{}, false
+		}
+		pluginData := *existing
+		pluginData.ReviewsCount = count
+		return pluginData, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !ok {
+		logger.Get(ctx).Debug("Failed to post the note: plugin data is missing")
+		return nil
+	}
+	ctx, _ = logger.WithField(ctx, "servicenow_incident_id", data.IncidentID)
+
+	slice := reqReviews[oldCount:]
+	if len(slice) == 0 {
+		return nil
+	}
+
+	errors := make([]error, 0, len(slice))
+	for _, review := range slice {
+		if err := a.servicenow.PostReviewNote(ctx, data.IncidentID, review); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// tryApproveRequest attempts to submit an approval if the requesting user is on-call in one of the services provided in request annotation.
+func (a *App) tryApproveRequest(ctx context.Context, req types.AccessRequest) error {
+	log := logger.Get(ctx)
+
+	serviceNames, err := a.getOnCallServiceNames(req)
+	if err != nil {
+		logger.Get(ctx).Debugf("Skipping the approval: %s", err)
+		return nil
+	}
+
+	onCallUsers := []string{}
+	for _, scheduleName := range serviceNames {
+		respondersResult, err := a.servicenow.GetOnCall(ctx, scheduleName)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		onCallUsers = append(onCallUsers, respondersResult...)
+	}
+
+	userIsOnCall := false
+	for _, user := range onCallUsers {
+		if req.GetUser() == user {
+			userIsOnCall = true
+		}
+	}
```

Comment:
Could we extract this to a function?
Also, if we invert the check for `if !userIsOnCall`, then we could remove one indentation level